### ISSUE TITLE
Sudden accuracy drop when training LSTM or GRU

### ISSUE THREAD

**Issue by marekmodry** [fchollet/keras#3425]
Hi! My **recurrent neural network (LSTM, resp. GRU)** behaves in a way I cannot explain. The training starts and it trains well (the results look quite good) when **suddenly accuracy drops (and loss rapidly increases)** - both training and testing metrics. Sometimes the net just _goes crazy_ and returns random outputs and sometimes (as in the last of three given examples) it starts to return _same output to all the inputs_. 

![image](https://cloud.githubusercontent.com/assets/8523511/17508785/f94aaf50-5e16-11e6-92cd-4d842b4c0e9c.png)

Do you have **any explanation for this behavior**? Any opinion is welcome. Please, see the task description and the figures below.

**The task:** From a word predict its word2vec vector
**The input:** We have an own word2vec model (normalized) and we feed the network with a word (letter by letter). We pad the words (see the example below).
**Example:** We have a word _football_ and we want to predict its word2vec vector which is 100 dimensions wide. Then the input is `$football$$$$$$$$$$`. 

Three examples of the behavior:

**Single layer LSTM**

```
model = Sequential([
    LSTM(1024, input_shape=encoder.shape, return_sequences=False),
    Dense(w2v_size, activation="linear")
])

model.compile(optimizer='adam', loss="mse", metrics=["accuracy"])
```

![image](https://cloud.githubusercontent.com/assets/8523511/17507845/43eecbfe-5e12-11e6-8ae3-179ac10d82b7.png)

**Single layer GRU**

```
model = Sequential([
    GRU(1024, input_shape=encoder.shape, return_sequences=False),
    Dense(w2v_size, activation="linear")
])

model.compile(optimizer='adam', loss="mse", metrics=["accuracy"])
```

![image](https://cloud.githubusercontent.com/assets/8523511/17507923/8ec75934-5e12-11e6-891b-a2d2f2c10e39.png)

**Double layer LSTM**

```
model = Sequential([
    LSTM(512, input_shape=encoder.shape, return_sequences=True),
    TimeDistributed(Dense(512, activation="sigmoid")),
    LSTM(512, return_sequences=False),
    Dense(256, activation="tanh"),
    Dense(w2v_size, activation="linear")
])

model.compile(optimizer='adam', loss="mse", metrics=["accuracy"])
```

![image](https://cloud.githubusercontent.com/assets/8523511/17507947/a29ad68e-5e12-11e6-9355-bde7711df71f.png)

We have also _experienced this kind of behavior in another project before_ which used similar architecture but its objective and data were different. Thus the reason should not be hidden in the data or in the particular objective but rather in the architecture.

**Comment by gcgibson1989** [comment]
My only 2 cents is that in the original W2V paper they used a shallow FFNN so there must be something about the RNN structures. I would think exploding/vanishing gradient but you are seeing it with the LSTM so....

**Comment by modrymar** [comment]
I have already experienced the same behavior with non-word2vec data and the behavior was very similar. Any idea on how to extract gradients from Keras? ... in order to examine closer whether the gradient is exploding/vanishing/... ?

**Comment by irfus** [comment]
I'm seeing something similar with a single layered LSTM classifier on acoustic data. After around 50 epochs, the loss suddenly started increasing and training and validation accuracy dropped badly.

```
modelV1 = Sequential()
modelV1.add(LSTM(60, return_sequences=False,
                     input_shape=(20, 3), activation='relu'))
modelV1.add(Dropout(0.2))
modelV1.add(Dense(7, activation='softmax'))
modelV1.compile(loss='categorical_crossentropy',
               optimizer='Nadam',
               metrics=['accuracy'])
modelV1.fit(F_train, V1_train, nb_epoch=100, validation_split=0.15)
```

**Comment by ylmeng** [comment]
Oh I had the same problem. Very annoying. I trained a network with LSTM and also used (word2vec), and got a sudden drop as in below:
Epoch 37/500
6400/6400 [==============================] - 44s - loss: 1.6209 - acc: 0.4892 - val_loss: 1.7309 - val_acc: 0.4562
Epoch 38/500
6400/6400 [==============================] - 44s - loss: 1.6402 - acc: 0.4805 - val_loss: 1.6928 - val_acc: 0.4463
Epoch 39/500
6400/6400 [==============================] - 44s - loss: 1.6328 - acc: 0.4917 - val_loss: 1.6404 - val_acc: 0.5044
Epoch 40/500
6400/6400 [==============================] - 44s - loss: 2.2682 - acc: 0.3142 - val_loss: 2.7337 - val_acc: 0.1806
Epoch 41/500
6400/6400 [==============================] - 44s - loss: 2.7492 - acc: 0.1811 - val_loss: 2.6720 - val_acc: 0.1806
Epoch 42/500
6400/6400 [==============================] - 44s - loss: 2.6774 - acc: 0.1814 - val_loss: 2.6636 - val_acc: 0.1806
Epoch 43/500
6400/6400 [==============================] - 44s - loss: 2.6739 - acc: 0.1814 - val_loss: 2.6631 - val_acc: 0.1806

**Comment by LetiP** [comment]
I have the same problem in a normal CNN using TensorFlow backend in keras. The loss frops to about 10^-7. Anybody an idea why this happens?

**Comment by pinkponk** [comment]
[accuracy_training_plot.pdf](https://github.com/fchollet/keras/files/857537/accuracy_training_plot.pdf)

Has this problem any solution yet!? It quite frustrating to have the training go bananas all of a sudden. 
I have run other CNN architectures on the same data but maybe it is some problem with sending in multiple references to the same data as I do? (I haven't tried inputting copies instead, training takes a day.)
Here is what I do:
---------------------------------------------------------------------------------------------------------------
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
....
history = model.fit([X_data_x,X_data_x,X_data_x, X_data_y,X_data_y,X_data_y], Y_data, validation_split=0.25, nb_epoch=epochs, batch_size=batch_size,callbacks=callbacks)

**Comment by thisismohitgupta** [comment]
It happens when somehow log returns NaN

**Comment by hongguangguo** [comment]
when I train model, using CNN with relu , acc and val_acc suddenly get zero.I'm clueless on what to do with this. It's bug? 
![screenshot from 2017-05-15 17-28-53](https://cloud.githubusercontent.com/assets/12960656/26051141/db8ed690-3993-11e7-8507-bd5786a53e68.png)

**Comment by ColaColin** [comment]
Switched from theano to Tensorflow today and saw this kind of thing happen for the first time ever.
Training went down to the expected best loss for the task at hand and then proceeded to suddenly drop to 15+ with an accuracy of less than 2%. It then stayed there, did not bounce back. Something must be numerically instable?

**Comment by wildanputra** [comment]
maybe because the relu,

The ReLU units can irreversibly die during training since they can get knocked off the data manifold. For example, you may find that as much as 40% of your network can be “dead” (i.e., neurons that never activate across the entire training dataset) if the learning rate is set too high. With a proper setting of the learning rate this is less frequently an issue. 
read more about it in here : http://lamda.nju.edu.cn/weixs/project/CNNTricks/CNNTricks.html

**Comment by marekmodry** [comment]
@wildanputra I don't believe that the dead relus would cause the network to drop the accuracy so suddenly. Moreover, in our initial case, there were no relus.

**Comment by MohammedSalman** [comment]
I am using Keras framework and this happened to me only when I am using SGD or nadam optimizer. It only works if I use adam optimizer. My model is 3 layers of LSTMs each with 100 cells.

**Comment by vmarkovtsev** [comment]
In my case, reducing the initial learning rate and adding decay helped.

**Comment by colpain** [comment]
This happened to me as well in two different RNN-based NLP tasks.
One is using LSTM to predict sentence topic, another one is to build GRU language model to predict next word.

![loss](https://user-images.githubusercontent.com/6015707/37241002-7ac83932-2407-11e8-85b4-9357e2971149.png)

any clue on this?

**Comment by free-variation** [comment]
Same here. NLP task with bidirectional LSTMs, using rmsprop and categorical_crossentropy.  ReLU units throughout.  Will try with `tanh` activations, but I've seen those are slow.

**Comment by GregVial** [comment]
I also faced this issue multiple times and it seems to me this is linked to vanishing gradient/dead neuron issue.
What helped (although it didn't solve the issue in all cases) was this:
- replace ReLU with SeLU
- check the input distribution and rescale it if it is too centered around 0 (you may want to add several standard deviations and/or multiply by a factor 2 or 3)

**Comment by Derise** [comment]
Same problem here. I've trained my model for about 5 times (with different labels) and it happend just the last time. It's a GRU-based model (tanh as activation) with ReLU. I'm not sure if that can be an issue but for the last time I've switched from tf 1.4 to 1.9. Look at last 15 minibatches during training:
Epoch 1/1
 - 4s - loss: 0.1395 - acc: 0.9561
Epoch 1/1
 - 4s - loss: 0.1351 - acc: 0.9542
Epoch 1/1
 - 4s - loss: 0.1436 - acc: 0.9538
Epoch 1/1
 - 4s - loss: 0.1385 - acc: 0.9555
Epoch 1/1
 - 4s - loss: 0.1379 - acc: 0.9547
Epoch 1/1
 - 4s - loss: 7.7643 - acc: 0.1823
Epoch 1/1
 - 4s - loss: 8.7400 - acc: 0.0702
Epoch 1/1
 - 4s - loss: 8.6492 - acc: 0.0771
Epoch 1/1
 - 4s - loss: 8.4718 - acc: 0.0842
Epoch 1/1
 - 4s - loss: 8.2733 - acc: 0.0865
Epoch 1/1
 - 1s - loss: 8.0101 - acc: 0.0988
Epoch 1/1
 - 6s - loss: 8.0152 - acc: 0.0924
Epoch 1/1
 - 6s - loss: 7.9908 - acc: 0.0947
Epoch 1/1
 - 6s - loss: 7.8206 - acc: 0.1009
Epoch 1/1
 - 6s - loss: 7.6227 - acc: 0.1053

**Comment by Dawei84** [comment]
For me, reducing the learning rate of ADAM from `lr=0.001` to `lr=0.0005` fixed it, using a single layer LSTM (ELU) for a classification problem.

**Comment by 200987299** [comment]
![Screenshot from 2021-01-09 19-28-01](https://user-images.githubusercontent.com/33377212/104090301-d3703500-52b0-11eb-80de-fcf3ed5bae31.png)
 I have a similar bug, any idea?
